### PR TITLE
feat: Add migrator and builder

### PR DESCRIPTION
**Pull request by keelerm84**

### CODE REVIEW

File: launchdarkly-server-sdk/src/migrations/migrator.rs [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+pub struct MigrationResultPair {
+    value: serde_json::Value,
+    origin: Origin,
+}
+
+/// MigrationResult represents the result of a migration operation. If the operation was
+/// successful, the result will contain a pair of values representing the result of the operation
+/// and the origin it was executed against. If the operation failed, the result will contain an
+/// error.
+type MigrationResult = Result<MigrationResultPair, Box<dyn std::error::Error>>;
```
Comment (cwaldren-ld) [inline launchdarkly-server-sdk/src/migrations/migrator.rs:19]:
```suggestion
/// MigrationResult represents the result of a migration operation. If the operation was
/// successful, the result will contain a [MigrationResultPair]. If the operation failed, the result will contain an error.
type MigrationResult = Result<MigrationResultPair, Box<dyn std::error::Error>>;
```

I think given the docs above on `MigrationResultPair`, it is slightly confusing to be mentioning "the result" twice - easier to just name the two parts of the variant and let people click over. 🤷‍♂️

File: launchdarkly-server-sdk/src/migrations/migrator.rs [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+pub struct MigrationWriteResult {
+    authoritative: MigrationResult,
+    nonauthoritative: Option<MigrationResult>,
+}
+
+// MigrationComparisonFn is used to compare the results of two migration operations. If the
+// provided results are equal, this method will return true and false otherwise.
+type MigrationComparisonFn = fn(serde_json::Value, serde_json::Value) -> bool;
+
+// MigrationImplFn represents the customer defined migration operation function. This method is
```
Comment (cwaldren-ld) [inline launchdarkly-server-sdk/src/migrations/migrator.rs:36]:
```suggestion
// MigrationImplFn represents the user defined migration operation function. This method is
```

(I see `user` used in other comments)

File: launchdarkly-server-sdk/src/migrations/migrator.rs [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+    pub fn track_errors(mut self, measure: bool) -> Self {
+        self.measure_errors = measure;
+        self
+    }
+
+    /// Read can be used to configure the migration-read behavior of the resulting
+    /// [Migrator] instance.
+    ///
+    /// Users are required to provide two different read methods -- one to read from the old migration origin, and one
+    /// to read from the new origin. Additionally, customers can opt-in to consistency tracking by providing a
```
Comment (cwaldren-ld) [inline launchdarkly-server-sdk/src/migrations/migrator.rs:119]:
```suggestion
    /// to read from the new origin. Additionally, users can opt-in to consistency tracking by providing a
```

File: launchdarkly-server-sdk/src/migrations/migrator.rs [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+        mut self,
+        old: MigrationImplFn,
+        new: MigrationImplFn,
+        compare: MigrationComparisonFn,
+    ) -> Self {
+        self.write_config = Some(MigrationConfig { old, new, compare });
+        self
+    }
+
+    // TODO: Do we make this a real error type?
```
Comment (cwaldren-ld) [inline launchdarkly-server-sdk/src/migrations/migrator.rs:150]: (TODO) 

Is the only error this read config error so far? 

Is there something actionable they could do in code to fix it?

File: launchdarkly-server-sdk/src/migrations/migrator.rs (hunk 1 above)
Comment (keelerm84) [reply launchdarkly-server-sdk/src/migrations/migrator.rs:19]: I end up reworking these structures in a subsequent PR. Is it okay if we merge this as is and then we make any future changes there?

The general shape as it stands on my latest local branch.

```rust
/// An internally used struct to represent the result of a migration operation along with the
/// origin it was executed against.
pub struct MigrationOriginResult {
    origin: Origin,
    result: MigrationResult,
}

// /// MigrationResult represents the result of a migration operation. If the operation was
// /// successful, the result will contain a pair of values representing the result of the operation
// /// and the origin it was executed against. If the operation failed, the result will contain an
// /// error.
type MigrationResult = Result<serde_json::Value, String>;
// TODO: Probably want to use a real error type here.

/// A write result contains the operation results against both the authoritative and
/// non-authoritative origins.
///
/// Authoritative writes are always executed first. In the event of a failure, the
/// non-authoritative write will not be executed, resulting in a `None` value in the final
/// MigrationWriteResult.
pub struct MigrationWriteResult {
    authoritative: MigrationOriginResult,
    nonauthoritative: Option<MigrationOriginResult>,
}
```

File: launchdarkly-server-sdk/src/migrations/migrator.rs (hunk 4 above)
Comment (keelerm84) [reply launchdarkly-server-sdk/src/migrations/migrator.rs:150]: These todos are for the length of the feature branch, so we don't have to block on this, but we should discuss it.

I don't think there is going to be much you can do programmatically. I mean, theoretically you could inspect the error and realize you didn't provide a read config and so you add one. But really, that's insane IMO which is why I just went with a string.

What do you think?

Comment (cwaldren-ld) [reply launchdarkly-server-sdk/src/migrations/migrator.rs:150]: Yea, I think that's fine.

**Review by cwaldren-ld — APPROVED** [review]

**Pull request merged**